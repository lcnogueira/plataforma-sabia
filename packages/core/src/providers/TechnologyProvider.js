import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { formatMoney } from '../utils';

export const TechnologyContext = React.createContext({});

export const TechnologyProvider = ({ children, technology }) => {
	const implementationCosts = useMemo(() => {
		const costs = technology?.technologyCosts?.costs?.implementation_costs;

		const total = costs?.reduce((acc, item) => acc + item?.quantity * item?.value, 0);

		return total ? formatMoney(total) : 0;
	}, [technology]);

	return (
		<TechnologyContext.Provider value={{ technology, implementationCosts }}>
			{children}
		</TechnologyContext.Provider>
	);
};

TechnologyProvider.propTypes = {
	children: PropTypes.node.isRequired,
	technology: PropTypes.shape({}),
};

TechnologyProvider.defaultProps = {
	technology: {},
};

export default TechnologyProvider;
