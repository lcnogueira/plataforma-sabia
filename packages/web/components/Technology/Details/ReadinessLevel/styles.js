import styled, { css } from 'styled-components';

export const Container = styled.div`
	${({ theme: { screens } }) => css`
		display: flex;
		position: relative;

		@media screen and (max-width: ${screens.medium}px) {
			> img {
				max-width: 100%;
			}
		}
	`}
`;

export const Marker = styled.div`
	position: absolute;
	right: 0;
	bottom: 0;
	width: 0;
	height: 0;
`;

export const MarkerBox = styled.div`
	${({ theme: { colors }, levelHeight, markerSize, currentLevel }) => css`
		min-width: 2rem;
		${Marker} {
			border-top: ${markerSize}px solid transparent;
			border-bottom: ${markerSize}px solid transparent;
			border-left: ${markerSize}px solid ${colors.primary};

			bottom: ${(currentLevel === 1 ? 0 : currentLevel - 1) * levelHeight}px;
			margin-right: calc(100% - ${markerSize}px);
		}
	`}
`;
