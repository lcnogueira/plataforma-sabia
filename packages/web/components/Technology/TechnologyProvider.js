import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { formatMoney } from '@sabia/core';
import TechnologyContext from './TechnologyContext';

export const TechnologyProvider = ({ children, technology }) => {
	const implementationCosts = useMemo(() => {
		const costs = technology?.technologyCosts?.costs?.implementation_costs;

		const total = costs?.reduce((acc, item) => acc + item?.quantity * item?.value, 0);

		return formatMoney(total);
	}, [technology]);

	return (
		<TechnologyContext.Provider value={{ technology, implementationCosts }}>
			{children}
		</TechnologyContext.Provider>
	);
};

TechnologyProvider.propTypes = {
	children: PropTypes.node.isRequired,
	technology: PropTypes.shape({}),
};

TechnologyProvider.defaultProps = {
	technology: {},
};

export default TechnologyProvider;
