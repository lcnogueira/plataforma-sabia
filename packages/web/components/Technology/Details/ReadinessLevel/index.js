import React, { useEffect, useRef, useState } from 'react';

import { useTechnology } from '../../../../hooks';
import { Container, MarkerBox, Marker } from './styles';

const LEVELS = 9;

const ReadinessLevel = () => {
	const imgRef = useRef();
	const [imageHeight, setImageHeight] = useState(0);
	const [levelHeight, setLevelHeight] = useState(0);

	const { technology: { taxonomies: { stage = '' } = {} } = {} } = useTechnology();
	const currentLevel = Number(stage.split(' ', 2)[1]);

	useEffect(() => {
		let timerId = null;
		const debouncedResize = () => {
			clearTimeout(timerId);
			timerId = setTimeout(() => setImageHeight(imgRef.current?.clientHeight), 250);
		};

		window.addEventListener('resize', debouncedResize);

		return () => {
			window.removeEventListener('resize', debouncedResize);
		};
	}, []);

	useEffect(() => {
		setLevelHeight(Math.floor(imageHeight / LEVELS));
	}, [imageHeight]);

	return (
		<Container>
			<MarkerBox
				levelHeight={levelHeight}
				markerSize={levelHeight / 2}
				imageHeight={imageHeight}
				currentLevel={currentLevel}
			>
				{!!currentLevel && !!levelHeight && !!imageHeight && <Marker />}
			</MarkerBox>
			<img
				ref={imgRef}
				src="/technology-readiness-level.svg"
				alt="Imagem representando a escala com os estÃ¡gios de desenvolvimento da tecnologia"
				onLoad={() => setImageHeight(imgRef.current?.clientHeight)}
			/>
		</Container>
	);
};

export default ReadinessLevel;
