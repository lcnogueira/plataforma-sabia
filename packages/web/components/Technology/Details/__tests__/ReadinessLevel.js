import React from 'react';
import { render, screen } from 'test-utils';
import { TechnologyProvider } from '../../TechnologyProvider';

import ReadinessLevel from '../ReadinessLevel';

const technologyMock = {
	terms: [
		{
			term:
				'Nível 4 - Verificação funcional de componente e/ou subsistema em ambiente laboratorial',
			slug:
				'nivel-4-verificacao-funcional-de-componente-e/ou-subsistema-em-ambiente-laboratorial',
			taxonomy: {
				taxonomy: 'STAGE',
			},
		},
	],
};

describe('<ReadinessLevel />', () => {
	it('should render without crashing', () => {
		const { container } = render(
			<TechnologyProvider technology={technologyMock}>
				<ReadinessLevel />
			</TechnologyProvider>,
		);

		expect(
			screen.getByRole('img', { src: '/technology-readiness-level.svg' }),
		).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
});
