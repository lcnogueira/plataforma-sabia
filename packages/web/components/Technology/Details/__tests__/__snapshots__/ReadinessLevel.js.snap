// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadinessLevel /> should render without crashing 1`] = `
.c2 {
  position: absolute;
  left: 0;
  bottom: 0;
  width: 0;
  height: 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  padding-left: 8px;
}

.c0 .c1 {
  border-top: 0px solid transparent;
  border-bottom: 0px solid transparent;
  border-left: 0px solid hsla(29,94%,62%);
  bottom: 0px;
}

@media screen and (max-width:768px) {
  .c0 > img {
    max-width: 100%;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 c2"
    />
    <img
      alt="Imagem representando a escala com os estÃ¡gios de desenvolvimento da tecnologia"
      src="/technology-readiness-level.svg"
    />
  </div>
</div>
`;
